File: src/lib/session.ts
import { getCookie } from "hono/cookie";
import { createMiddleware } from "hono/factory";

type Env = {
  Bindings: { DB: D1Database; SESSION_SECRET: string };
  Variables: { session: SessionPayload };
};

const COOKIE_NAME = "session";
const SESSION_TTL = 7 * 86400;
const ALGO = { name: "HMAC", hash: "SHA-256" };

export type SessionPayload = {
  sub: string;
  email: string;
  name: string;
  iat: number;
  exp: number;
};

async function hmacSign(data: string, secret: string): Promise<string> {
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey("raw", enc.encode(secret), ALGO, false, ["sign"]);
  const sig = await crypto.subtle.sign(ALGO, key, enc.encode(data));
  return btoa(String.fromCharCode(...new Uint8Array(sig)))
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

async function hmacVerify(data: string, sigB64: string, secret: string): Promise<boolean> {
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey("raw", enc.encode(secret), ALGO, false, ["verify"]);
  const sig = Uint8Array.from(atob(sigB64.replace(/-/g, "+").replace(/_/g, "/")), (c) =>
    c.charCodeAt(0),
  );
  return crypto.subtle.verify(ALGO, key, sig, enc.encode(data));
}

function toB64(obj: unknown): string {
  return btoa(JSON.stringify(obj)).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function fromB64(b64: string): unknown {
  return JSON.parse(atob(b64.replace(/-/g, "+").replace(/_/g, "/")));
}

export async function signToken(payload: SessionPayload, secret: string): Promise<string> {
  const encoded = toB64(payload);
  const sig = await hmacSign(encoded, secret);
  return `${encoded}.${sig}`;
}

export async function verifyToken(token: string, secret: string): Promise<SessionPayload | null> {
  const dot = token.lastIndexOf(".");
  if (dot === -1) return null;
  const encoded = token.slice(0, dot);
  const sig = token.slice(dot + 1);
  const ok = await hmacVerify(encoded, sig, secret);
  if (!ok) return null;
  const payload = fromB64(encoded) as SessionPayload;
  if (typeof payload !== "object" || !payload?.sub) return null;
  return payload;
}

function makePayload(sub: string, email: string, name: string): SessionPayload {
  const now = Math.floor(Date.now() / 1000);
  return { sub, email, name, iat: now, exp: now + SESSION_TTL };
}

export { makePayload };

export const session = () =>
  createMiddleware<Env>(async (c, next) => {
    const secret = c.env.SESSION_SECRET || "insecure-dev-default";
    const raw = getCookie(c, COOKIE_NAME);
    if (raw) {
      try {
        const payload = await verifyToken(raw, secret);
        if (payload) c.set("session", payload);
      } catch {
        // ignore
      }
    }
    await next();
  });

export const requireSession = createMiddleware<Env>(async (c, next) => {
  const session = c.get("session");
  if (!session) {
    if (c.req.header("HX-Request")) {
      c.header("HX-Redirect", "/auth/login");
      return c.text("Unauthorized", 401);
    }
    return c.redirect("/auth/login");
  }
  await next();
});

File: src/lib/session.py
import hmac
import hashlib
import base64
import json
import time
from fastapi import Request, HTTPException


COOKIE_NAME = "session"
SESSION_TTL = 7 * 86400


def to_b64(data: dict) -> str:
    payload = json.dumps(data, separators=(",", ":"))
    return base64.urlsafe_b64encode(payload.encode()).rstrip(b"=").decode()


def from_b64(raw: str) -> dict | None:
    padded = raw + "=" * (4 - len(raw) % 4)
    try:
        return json.loads(base64.urlsafe_b64decode(padded))
    except Exception:
        return None


def sign_token(payload: dict, secret: str) -> str:
    encoded = to_b64(payload)
    sig = hmac.new(secret.encode(), encoded.encode(), hashlib.sha256).digest()
    sig_b64 = base64.urlsafe_b64encode(sig).rstrip(b"=").decode()
    return f"{encoded}.{sig_b64}"


def verify_token(token: str, secret: str) -> dict | None:
    parts = token.rsplit(".", 1)
    if len(parts) != 2:
        return None
    encoded, sig_b64 = parts
    try:
        sig = base64.urlsafe_b64decode(sig_b64 + "=" * (4 - len(sig_b64) % 4))
    except Exception:
        return None
    expected = hmac.new(secret.encode(), encoded.encode(), hashlib.sha256).digest()
    if not hmac.compare_digest(sig, expected):
        return None
    payload = from_b64(encoded)
    if not payload or "sub" not in payload:
        return None
    if payload.get("exp", 0) < time.time():
        return None
    return payload


def make_payload(sub: str, email: str, name: str) -> dict:
    now = int(time.time())
    return {"sub": sub, "email": email, "name": name, "iat": now, "exp": now + SESSION_TTL}


def _get_secret(request: Request) -> str:
    try:
        return request.scope["env"].SESSION_SECRET or "insecure-dev-default"
    except AttributeError:
        return "insecure-dev-default"
    except KeyError:
        return "insecure-dev-default"


async def session_middleware(request: Request, call_next):
    raw = request.cookies.get(COOKIE_NAME)
    session = None
    if raw:
        secret = _get_secret(request)
        session = verify_token(raw, secret)
    request.state.session = session
    response = await call_next(request)
    return response


async def require_session(request: Request) -> dict:
    session = getattr(request.state, "session", None)
    if not session:
        if request.headers.get("HX-Request") == "true":
            raise HTTPException(
                status_code=401,
                detail="Unauthorized",
                headers={"HX-Redirect": "/auth/login"},
            )
        raise HTTPException(status_code=302, headers={"Location": "/auth/login"})
    return session
